File: Prácticas/P5/JAVA/Civitas/src/civitas/TituloPropiedad.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

/**
 *
 * @author Jose Padial
 */
public class TituloPropiedad {
    private float alquilerBase;
    private static final float factorInteresHipoteca = (float) 1.1;
    private float factorRevalorizacion;
    private float hipotecaBase;
    private boolean hipotecado;
    private String nombre;
    private int numCasas;
    private int numHoteles;
    private float precioCompra;
    private float precioEdificar;
    private Jugador propietario;
    
    TituloPropiedad(String nomb, float ab, float fr, float hb, float pc, float pe){
        nombre = nomb;
        alquilerBase = ab;
        factorRevalorizacion = fr;
        hipotecaBase = hb;
        precioCompra = pc;
        precioEdificar = pe;
        hipotecado = false;
        numCasas = 0;
        numHoteles = 0;
        propietario = null;
    }
    
    void actualizarPropietarioPorConversion(Jugador jugador){
        propietario = jugador;
    }
    
    boolean cancelarHipoteca(Jugador jugador){
        boolean resultado = false;
        if(hipotecado == true && esEsteElPropietario(jugador) == true){
            jugador.paga(getImporteCancelarHipoteca());
            resultado = true;
            hipotecado = false;
        }
        return resultado;
    }
    
    int cantidadCasasHoteles(){
        return (numHoteles + numCasas);
    }
    
    boolean comprar(Jugador jugador){
        boolean resultado = false;
        if(propietario == null){
            propietario = jugador;
            resultado = true;
            jugador.paga(precioCompra);
        }
        return resultado;
    }
    
    boolean construirCasa(Jugador jugador){
        boolean resultado = false;
        if(esEsteElPropietario(jugador) == true){
            jugador.paga(precioEdificar);
            numCasas= numCasas+1;
            resultado = true;
        }
        return resultado;
    }
    
    boolean construirHotel(Jugador jugador){
        boolean resultado = false;
        if(esEsteElPropietario(jugador) == true){
            jugador.paga(precioEdificar);
            numHoteles= numHoteles+1;
            resultado = true;
        }
        return resultado;
    }
    
    boolean derruirCasas(int n, Jugador jugador){
        if(jugador == propietario && numCasas >= n){
            numCasas = numCasas - n;
            return true;
        }
        else{
            return false;
        }
    }
    
    private boolean esEsteElPropietario(Jugador jugador){
        if(jugador == propietario){
            return true;
        }
        else{
            return false;
        }
    }
    
    public boolean getHipotecado(){
        return hipotecado;
    }
    
    float getImporteCancelarHipoteca(){
        return (hipotecaBase * factorInteresHipoteca);
        
    }
    
    float getImporteHipoteca(){
        return hipotecaBase;
    }
    
    public String getNombre(){
        return nombre;
    }
    
    public int getNumCasas(){
        return numCasas;
    }
    
    public int getNumHoteles(){
        return numHoteles;
    }
    
    public  float getPrecioAlquiler(){
        
        if(hipotecado == true || propietario.getEncarcelado()){
            return 0;
        }
        else{
            return alquilerBase;
        }
    }
    
    public float getPrecioCompra(){
        return precioCompra;
    }
    
    float getPrecioEdificar(){
        return precioEdificar;
    }
    
    private float getPrecioVenta(){
        float importe = 0;
        float precioHoteles = numHoteles * precioEdificar;
        float precioCasas = numCasas * precioEdificar;
        
        importe = (precioCasas + precioHoteles) * factorRevalorizacion;
        importe = precioCompra + importe;
        
        return importe;
    }
    
    Jugador getPropietario(){
        return propietario;
    }
    
    boolean hipotecar(Jugador jugador){
        boolean salida = false;
        if(hipotecado == false && esEsteElPropietario(jugador)==true){
            jugador.recibe(getImporteHipoteca());
            hipotecado = true;
            salida = true;
        }
        return salida;
    }
    
    private boolean propietarioEncarcelado(){
        if(propietario.getEncarcelado() == true){
            return true;
        }
        else{
            return false;
        }
    }
    
    boolean tienePropietario(){
        if(propietario != null){
            return true;
        }
        else{
            return false;
        }
    }
    
    void tramitarAlquiler(Jugador jugador){
        if(propietario != null && jugador != propietario){
            jugador.paga(getPrecioAlquiler());
            propietario.recibe(getPrecioAlquiler());
        }
    }
    
    boolean vender(Jugador jugador){
        if(jugador == propietario && hipotecado == false){
            jugador.recibe(getPrecioVenta());
            propietario = null;
            numCasas = 0;
            numHoteles = 0;
            return true;
        }
        else{
            return false;
        }
    }
    
    @Override
    public String toString() {
        return "TituloPropiedad{" + "nombre=" + nombre + ", hipotecada=" + hipotecado + ", numHoteles=" + numHoteles + ", numCasas=" + numCasas + '}';
    }
 
}

File: Prácticas/P4/RUBY/Civitas/lib/titulo_propiedad.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative"jugador.rb"
require_relative"diario.rb"

module Civitas
  
class Titulo_propiedad
  
  @@FACTOR_INTERESES_HIPOTECA = 1.1
  
  def initialize(nombre,ab,fr,hb,pc,pe)
    @nombre = nombre
    @alquiler_base = ab
    @factor_revalorizacion = fr
    @hipoteca_base = hb
    @precio_compra = pc
    @precio_edificar = pe
    @hipotecado = false
    @num_casas = 0
    @num_hoteles = 0
    @propietario = nil
  end
  
  def actualiza_propietario_por_conversion(jugador)
    @propietario = jugador
  end
  
  def cancelar_hipoteca(jugador)
    resultado = false
    if(@hipotecado == true && es_este_el_propietario(jugador) == true)
      jugador.paga(get_importe_cancelar_hipoteca)
      resultado = true
      @hipotecado = false
    end
        return resultado
  end
  
  def cantidad_casas_hoteles
    return (@num_casas + @num_hoteles)
  end

  def comprar(jugador)
    resultado = false
    if(@propietario == nil)
      @propietario = jugador
      resultado = true
      jugador.paga(@precio_compra)
    end
        return resultado
  end

  def construir_casa(jugador)
    resultado = false
    if(es_este_el_propietario(jugador) == true)
        jugador.paga(@precio_edificar)
        @num_casas = @num_casas+1
        resultado = true
    end
        return resultado
  end
  
  def construir_hotel(jugador)
    resultado = false
    if(es_este_el_propietario(jugador) == true)
        jugador.paga(@precio_edificar)
        @num_hoteles = @num_hoteles + 1
        resultado = true
    end
        return resultado
  end

  def derruir_casas(n,jugador)
    if(jugador == @propietario && @num_casas >= n)
        @num_casas = @num_casas - n
        return true
    else
        return false
    end 
  end
  
  def get_hipotecado
    @hipotecado
  end
  
  def get_importe_cancelar_hipoteca
    return (@hipoteca_base * @@FACTOR_INTERESES_HIPOTECA)
  end
  
  def get_nombre
    @nombre
  end
  
  def get_num_casas
    @num_casas
  end
  
  def get_num_hoteles
    @num_hoteles
  end
 
  def get_precio_compra
    @precio_compra
  end
  
  def get_precio_edificar
    @precio_edificar
  end
  
  def get_propietario
    @propietario
  end
  
  def hipotecar(jugador)
     salida = false
     if(@hipotecado == false && es_este_el_propietario(jugador)==true)
         jugador.recibe(get_importe_hipoteca)
         @hipotecado = true
         salida = true
     end
        return salida
  end
  
  def tiene_propietario
    if(@propietario != nil)
        return true
    else
        return false
    end   
  end
  
  def tramitar_alquiler(jugador)
    if(@propietario != nil && jugador != @propietario)
        jugador.paga(get_precio_alquiler)
        @propietario.recibe(get_precio_alquiler)
    end
  end
  
  def vender(jugador)
    if(jugador == @propietario && @hipotecado == false)
        jugador.recibe(get_precio_venta)
        @propietario = nil
        @num_casas = 0
        @num_hoteles = 0
        return true
    else
        return false
    end
  end
  
  private
  def propietario_encarcelado
    if(@propietario.get_encarcelado == true)
      return true
    else
      return false
    end
  end
  
  def get_precio_venta
     precioHoteles = @num_hoteles * @precio_edificar
     precioCasas = @num_casas * @precio_edificar
     importe = (precioCasas + precioHoteles) * @factor_revalorizacion
     importe = @precio_compra + importe
     return importe
  end
  
  def get_precio_alquiler
    if(@hipotecado == true || propietario_encarcelado == true)
      return 0
    else
      @alquiler_base
    end
  end
  
  def get_importe_hipoteca
    @hipoteca_base
  end
  
  def es_este_el_propietario(jugador)
    if(jugador == @propietario)
        return true
    else
        return false
    end  
  end
  
  public
  def to_s
      to_s = "Nombre: #{@nombre} Hipotecada: #{@hipotecada} Precio de compra: #{@precio_compra} Alquiler base: #{@alquiler_base} Factor de revalorizacion: #{@factor_revalorizacion}% Hipoteca base: #{@hipoteca_base} Precio de edificacion: #{@precio_edificar} Numero de casas: #{@num_casas} Numero de hoteles: #{@num_hoteles}\n"
  end
  
end

end
